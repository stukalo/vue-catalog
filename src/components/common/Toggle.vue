<template>
  <div class='toggle'>
    <div class='toggle_title'>
      <span>{{ title }}</span>
    </div>
    <ul class="toggle_options options">
      <li class='options_item'
          v-bind:class="option.active && 'active'"
          v-for="option in options"
          :key="option.value"
          @click="this.onChange(option.value)"
      >
        <span>
          {{ option.text }}
        </span>
      </li>
    </ul>
  </div>
</template>

<script>
export default {
  name: 'Toggle',
  props: ['title', 'options'],
  methods: {
    onChange(value) {
      this.$emit('change', value);
    },
  },
};
</script>

<style scoped lang="less">
@import (reference) '../../assets/css/variables.less';
@import (once) '../../assets/css/reset.less';

.toggle {
  display: flex;
  width: fit-content;
  align-items: center;
  .fontMixin(18px);
  color: @primaryText;
  text-transform: uppercase;
}

.toggle_title {
  margin-right: 20px;
}

.toggle_options {
  display: flex;
  list-style: none;
  background-color: @secondaryBackgroundOpacity;
  border-radius: @primaryBorderRadius;
  overflow: hidden;
}

.options_item {
  padding: 0 20px;
  height: 40px;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  background-color: @primaryBackgroundOpacity;

  &.active {
    background-color: @activeBackground;
  }
}

</style>
