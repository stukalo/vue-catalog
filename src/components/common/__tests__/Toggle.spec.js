import { shallowMount } from '@vue/test-utils';
import Toggle from '../Toggle.vue';

const props = {
  title: 'Title',
  options: [{
    text: 'Option-1',
    value: 'value-1',
    active: false,
  }, {
    text: 'Option-2',
    value: 'value-2',
    active: false,
  }]
};

describe('Toggle', () => {
  describe('render', () => {
    it('should render correctly', () => {
      const wrapper = shallowMount(Toggle);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallowMount(Toggle, {
        propsData: props,
      });
    });

    it('should have correct title', () => {
      const title = wrapper.find('.toggle_title>span');
      expect(title.text()).toEqual(props.title);
    });

    it('should have correct options and emmit change event', () => {
      const optionsItems = wrapper.findAll('.options_item');
      props.options.forEach(({ value, text, active }, index) => {
        const item = optionsItems[index];
        item.trigger('click');

        expect(wrapper.emitted().change[index][0]).toEqual(value);
        expect(item.find('span').text()).toEqual(text);
        expect(item.classes().includes('active')).toEqual(active);
      });
    });
  });
});
