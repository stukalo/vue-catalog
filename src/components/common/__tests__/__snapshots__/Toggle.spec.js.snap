// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle render should render correctly 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <div
        class="toggle"
      >
        <div
          class="toggle_title"
        >
          <span />
        </div>
        <ul
          class="toggle_options options"
        >
          
          
        </ul>
      </div>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
        },
        "transition-group": Object {
          "name": "transition-group",
        },
      },
      "config": Object {
        "compilerOptions": Object {},
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "emitsCache": WeakMap {},
      "mixins": Array [
        Object {
          "beforeCreate": [Function],
        },
      ],
      "optionsCache": WeakMap {},
      "propsCache": WeakMap {},
      "provides": Object {},
      "reload": [Function],
    },
    "_instance": Object {
      "a": null,
      "accessCache": Object {},
      "appContext": Object {
        "app": [Circular],
        "components": Object {
          "transition": Object {
            "name": "transition",
          },
          "transition-group": Object {
            "name": "transition-group",
          },
        },
        "config": Object {
          "compilerOptions": Object {},
          "errorHandler": undefined,
          "globalProperties": Object {},
          "isNativeTag": [Function],
          "optionMergeStrategies": Object {},
          "performance": false,
          "warnHandler": undefined,
        },
        "directives": Object {},
        "emitsCache": WeakMap {},
        "mixins": Array [
          Object {
            "beforeCreate": [Function],
          },
        ],
        "optionsCache": WeakMap {},
        "propsCache": WeakMap {},
        "provides": Object {},
        "reload": [Function],
      },
      "asyncDep": null,
      "asyncResolved": false,
      "attrs": Object {},
      "bc": null,
      "bm": null,
      "bu": null,
      "bum": null,
      "c": null,
      "components": null,
      "ctx": Object {},
      "da": null,
      "data": Object {},
      "directives": null,
      "ec": null,
      "emit": [Function],
      "emitsOptions": null,
      "emitted": null,
      "exposeProxy": null,
      "exposed": null,
      "inheritAttrs": undefined,
      "isDeactivated": false,
      "isMounted": true,
      "isUnmounted": false,
      "m": null,
      "next": null,
      "parent": null,
      "props": Object {},
      "propsDefaults": Object {},
      "propsOptions": Array [
        Object {},
        Array [],
      ],
      "provides": Object {},
      "proxy": Object {},
      "refs": Object {
        "VTU_COMPONENT": Object {
          "hasOwnProperty": [Function],
          "onChange": [Function],
          "options": Array [],
          "title": "",
        },
      },
      "render": [Function],
      "renderCache": Array [],
      "root": [Circular],
      "rtc": null,
      "rtg": null,
      "scope": EffectScope {
        "active": true,
        "cleanups": Array [],
        "effects": Array [
          ReactiveEffect {
            "active": true,
            "allowRecurse": true,
            "deps": Array [
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
            ],
            "fn": [Function],
            "onTrack": undefined,
            "onTrigger": undefined,
            "scheduler": [Function],
          },
        ],
      },
      "setupContext": null,
      "setupState": Object {},
      "slots": Object {},
      "sp": null,
      "subTree": Object {
        "__v_isVNode": true,
        "__v_skip": true,
        "anchor": null,
        "appContext": null,
        "children": null,
        "component": Object {
          "a": null,
          "accessCache": Object {
            "@@__IMMUTABLE_ITERABLE__@@": 0,
            "@@__IMMUTABLE_RECORD__@@": 0,
            "_isMockFunction": 0,
            "constructor": 0,
            "hasOwnProperty": 4,
            "isVueInstance": 0,
            "onChange": 4,
            "options": 3,
            "title": 3,
            "toJSON": 0,
            Symbol(Symbol.toStringTag): 0,
          },
          "appContext": Object {
            "app": [Circular],
            "components": Object {
              "transition": Object {
                "name": "transition",
              },
              "transition-group": Object {
                "name": "transition-group",
              },
            },
            "config": Object {
              "compilerOptions": Object {},
              "errorHandler": undefined,
              "globalProperties": Object {},
              "isNativeTag": [Function],
              "optionMergeStrategies": Object {},
              "performance": false,
              "warnHandler": undefined,
            },
            "directives": Object {},
            "emitsCache": WeakMap {},
            "mixins": Array [
              Object {
                "beforeCreate": [Function],
              },
            ],
            "optionsCache": WeakMap {},
            "propsCache": WeakMap {},
            "provides": Object {},
            "reload": [Function],
          },
          "asyncDep": null,
          "asyncResolved": false,
          "attrs": Object {},
          "bc": null,
          "bm": null,
          "bu": null,
          "bum": null,
          "c": null,
          "components": Object {},
          "ctx": Object {
            "hasOwnProperty": [Function],
            "onChange": [Function],
            "options": Array [],
            "title": "",
          },
          "da": null,
          "data": Object {},
          "directives": null,
          "ec": null,
          "emit": [Function],
          "emitsOptions": null,
          "emitted": null,
          "exposeProxy": null,
          "exposed": null,
          "inheritAttrs": undefined,
          "isDeactivated": false,
          "isMounted": true,
          "isUnmounted": false,
          "m": null,
          "next": null,
          "parent": [Circular],
          "props": Object {
            "options": Array [],
            "title": "",
          },
          "propsDefaults": Object {},
          "propsOptions": Array [
            Object {
              "options": Object {
                "0": false,
                "1": true,
                "default": Array [],
                "type": [Function],
              },
              "title": Object {
                "0": false,
                "1": true,
                "default": "",
                "type": [Function],
              },
            },
            Array [
              "title",
              "options",
            ],
          ],
          "provides": Object {},
          "proxy": Object {
            "hasOwnProperty": [Function],
            "onChange": [Function],
            "options": Array [],
            "title": "",
          },
          "refs": Object {},
          "render": [Function],
          "renderCache": Array [],
          "root": [Circular],
          "rtc": null,
          "rtg": null,
          "scope": EffectScope {
            "active": true,
            "cleanups": Array [],
            "effects": Array [
              ReactiveEffect {
                "active": true,
                "allowRecurse": true,
                "deps": Array [
                  Set {
                    [Circular],
                  },
                  Set {
                    [Circular],
                  },
                ],
                "fn": [Function],
                "onTrack": undefined,
                "onTrigger": undefined,
                "scheduler": [Function],
              },
            ],
          },
          "setupContext": null,
          "setupState": Object {},
          "slots": Object {},
          "sp": null,
          "subTree": Object {
            "__v_isVNode": true,
            "__v_skip": true,
            "anchor": null,
            "appContext": null,
            "children": Array [
              Object {
                "__v_isVNode": true,
                "__v_skip": true,
                "anchor": null,
                "appContext": null,
                "children": Array [
                  Object {
                    "__v_isVNode": true,
                    "__v_skip": true,
                    "anchor": null,
                    "appContext": null,
                    "children": "",
                    "component": null,
                    "dirs": null,
                    "dynamicChildren": null,
                    "dynamicProps": null,
                    "el": <span />,
                    "key": null,
                    "patchFlag": 1,
                    "props": null,
                    "ref": null,
                    "scopeId": null,
                    "shapeFlag": 1,
                    "slotScopeIds": null,
                    "ssContent": null,
                    "ssFallback": null,
                    "staticCount": 0,
                    "suspense": null,
                    "target": null,
                    "targetAnchor": null,
                    "transition": null,
                    "type": "span",
                  },
                ],
                "component": null,
                "dirs": null,
                "dynamicChildren": null,
                "dynamicProps": null,
                "el": <div
                  class="toggle_title"
                >
                  <span />
                </div>,
                "key": null,
                "patchFlag": 0,
                "props": Object {
                  "class": "toggle_title",
                },
                "ref": null,
                "scopeId": null,
                "shapeFlag": 17,
                "slotScopeIds": null,
                "ssContent": null,
                "ssFallback": null,
                "staticCount": 0,
                "suspense": null,
                "target": null,
                "targetAnchor": null,
                "transition": null,
                "type": "div",
              },
              Object {
                "__v_isVNode": true,
                "__v_skip": true,
                "anchor": null,
                "appContext": null,
                "children": Array [
                  Object {
                    "__v_isVNode": true,
                    "__v_skip": true,
                    "anchor": ,
                    "appContext": null,
                    "children": Array [],
                    "component": null,
                    "dirs": null,
                    "dynamicChildren": Array [],
                    "dynamicProps": null,
                    "el": ,
                    "key": null,
                    "patchFlag": 128,
                    "props": null,
                    "ref": null,
                    "scopeId": null,
                    "shapeFlag": 16,
                    "slotScopeIds": null,
                    "ssContent": null,
                    "ssFallback": null,
                    "staticCount": 0,
                    "suspense": null,
                    "target": null,
                    "targetAnchor": null,
                    "transition": null,
                    "type": Symbol(Fragment),
                  },
                ],
                "component": null,
                "dirs": null,
                "dynamicChildren": null,
                "dynamicProps": null,
                "el": <ul
                  class="toggle_options options"
                >
                  
                  
                </ul>,
                "key": null,
                "patchFlag": 0,
                "props": Object {
                  "class": "toggle_options options",
                },
                "ref": null,
                "scopeId": null,
                "shapeFlag": 17,
                "slotScopeIds": null,
                "ssContent": null,
                "ssFallback": null,
                "staticCount": 0,
                "suspense": null,
                "target": null,
                "targetAnchor": null,
                "transition": null,
                "type": "ul",
              },
            ],
            "component": null,
            "dirs": null,
            "dynamicChildren": Array [
              Object {
                "__v_isVNode": true,
                "__v_skip": true,
                "anchor": null,
                "appContext": null,
                "children": "",
                "component": null,
                "dirs": null,
                "dynamicChildren": null,
                "dynamicProps": null,
                "el": <span />,
                "key": null,
                "patchFlag": 1,
                "props": null,
                "ref": null,
                "scopeId": null,
                "shapeFlag": 1,
                "slotScopeIds": null,
                "ssContent": null,
                "ssFallback": null,
                "staticCount": 0,
                "suspense": null,
                "target": null,
                "targetAnchor": null,
                "transition": null,
                "type": "span",
              },
              Object {
                "__v_isVNode": true,
                "__v_skip": true,
                "anchor": ,
                "appContext": null,
                "children": Array [],
                "component": null,
                "dirs": null,
                "dynamicChildren": Array [],
                "dynamicProps": null,
                "el": ,
                "key": null,
                "patchFlag": 128,
                "props": null,
                "ref": null,
                "scopeId": null,
                "shapeFlag": 16,
                "slotScopeIds": null,
                "ssContent": null,
                "ssFallback": null,
                "staticCount": 0,
                "suspense": null,
                "target": null,
                "targetAnchor": null,
                "transition": null,
                "type": Symbol(Fragment),
              },
            ],
            "dynamicProps": null,
            "el": <div
              class="toggle"
            >
              <div
                class="toggle_title"
              >
                <span />
              </div>
              <ul
                class="toggle_options options"
              >
                
                
              </ul>
            </div>,
            "key": null,
            "patchFlag": 0,
            "props": Object {
              "class": "toggle",
            },
            "ref": null,
            "scopeId": null,
            "shapeFlag": 17,
            "slotScopeIds": null,
            "ssContent": null,
            "ssFallback": null,
            "staticCount": 0,
            "suspense": null,
            "target": null,
            "targetAnchor": null,
            "transition": null,
            "type": "div",
          },
          "suspense": null,
          "suspenseId": 0,
          "type": Object {
            "components": Object {},
            "methods": Object {
              "onChange": [Function],
            },
            "name": "Toggle",
            "props": Object {
              "options": Object {
                "0": false,
                "1": true,
                "default": Array [],
                "type": [Function],
              },
              "title": Object {
                "0": false,
                "1": true,
                "default": "",
                "type": [Function],
              },
            },
            "render": [Function],
          },
          "u": null,
          "uid": 1,
          "um": null,
          "update": [Function],
          "vnode": [Circular],
          "withProxy": null,
        },
        "dirs": null,
        "dynamicChildren": null,
        "dynamicProps": null,
        "el": <div
          class="toggle"
        >
          <div
            class="toggle_title"
          >
            <span />
          </div>
          <ul
            class="toggle_options options"
          >
            
            
          </ul>
        </div>,
        "key": null,
        "patchFlag": 0,
        "props": Object {
          "ref": "VTU_COMPONENT",
        },
        "ref": Object {
          "i": [Circular],
          "r": "VTU_COMPONENT",
        },
        "scopeId": null,
        "shapeFlag": 4,
        "slotScopeIds": null,
        "ssContent": null,
        "ssFallback": null,
        "staticCount": 0,
        "suspense": null,
        "target": null,
        "targetAnchor": null,
        "transition": null,
        "type": Object {
          "components": Object {},
          "methods": Object {
            "onChange": [Function],
          },
          "name": "Toggle",
          "props": Object {
            "options": Object {
              "0": false,
              "1": true,
              "default": Array [],
              "type": [Function],
            },
            "title": Object {
              "0": false,
              "1": true,
              "default": "",
              "type": [Function],
            },
          },
          "render": [Function],
        },
      },
      "suspense": null,
      "suspenseId": 0,
      "type": Object {
        "name": "VTU_ROOT",
        "render": [Function],
      },
      "u": null,
      "uid": 0,
      "um": null,
      "update": [Function],
      "vnode": Object {
        "__v_isVNode": true,
        "__v_skip": true,
        "anchor": null,
        "appContext": Object {
          "app": [Circular],
          "components": Object {
            "transition": Object {
              "name": "transition",
            },
            "transition-group": Object {
              "name": "transition-group",
            },
          },
          "config": Object {
            "compilerOptions": Object {},
            "errorHandler": undefined,
            "globalProperties": Object {},
            "isNativeTag": [Function],
            "optionMergeStrategies": Object {},
            "performance": false,
            "warnHandler": undefined,
          },
          "directives": Object {},
          "emitsCache": WeakMap {},
          "mixins": Array [
            Object {
              "beforeCreate": [Function],
            },
          ],
          "optionsCache": WeakMap {},
          "propsCache": WeakMap {},
          "provides": Object {},
          "reload": [Function],
        },
        "children": null,
        "component": [Circular],
        "dirs": null,
        "dynamicChildren": null,
        "dynamicProps": null,
        "el": <div
          class="toggle"
        >
          <div
            class="toggle_title"
          >
            <span />
          </div>
          <ul
            class="toggle_options options"
          >
            
            
          </ul>
        </div>,
        "key": null,
        "patchFlag": 0,
        "props": null,
        "ref": null,
        "scopeId": null,
        "shapeFlag": 4,
        "slotScopeIds": null,
        "ssContent": null,
        "ssFallback": null,
        "staticCount": 0,
        "suspense": null,
        "target": null,
        "targetAnchor": null,
        "transition": null,
        "type": Object {
          "name": "VTU_ROOT",
          "render": [Function],
        },
      },
      "withProxy": null,
    },
    "_props": null,
    "_uid": 0,
    "component": [Function],
    "config": Object {
      "compilerOptions": Object {},
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.2.23",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "hasOwnProperty": [Function],
    "onChange": [Function],
    "options": Array [],
    "title": "",
  },
  "isDisabled": [Function],
  "rootVM": Object {},
  "wrapperElement": <div
    class="toggle"
  >
    <div
      class="toggle_title"
    >
      <span />
    </div>
    <ul
      class="toggle_options options"
    >
      
      
    </ul>
  </div>,
}
`;
